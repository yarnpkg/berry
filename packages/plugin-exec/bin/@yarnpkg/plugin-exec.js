/* eslint-disable */
module.exports = {
name: "@yarnpkg/plugin-exec",
factory: function (require) {
var plugin;plugin=(()=>{"use strict";var e={230:(e,t,r)=>{r.r(t),r.d(t,{execUtils:()=>a,default:()=>p});var a={};r.r(a),r.d(a,{loadGeneratorFile:()=>l,makeLocator:()=>c,makeSpec:()=>i,parseSpec:()=>s});var o=r(966),n=r(688);function s(e){const{params:t,selector:r}=o.structUtils.parseRange(e),a=n.npath.toPortablePath(r);return{parentLocator:t&&"string"==typeof t.locator?o.structUtils.parseLocator(t.locator):null,path:a}}function i({parentLocator:e,path:t,generatorHash:r,protocol:a}){const n=null!==e?{locator:o.structUtils.stringifyLocator(e)}:{},s=void 0!==r?{hash:r}:{};return o.structUtils.makeRange({protocol:a,source:t,selector:t,params:{...s,...n}})}function c(e,{parentLocator:t,path:r,generatorHash:a,protocol:n}){return o.structUtils.makeLocator(e,i({parentLocator:t,path:r,generatorHash:a,protocol:n}))}async function l(e,t,r){const{parentLocator:a,path:s}=o.structUtils.parseFileStyleRange(e,{protocol:t}),i=n.ppath.isAbsolute(s)?{packageFs:new n.CwdFS(n.PortablePath.root),prefixPath:n.PortablePath.dot,localPath:n.PortablePath.root}:await r.fetcher.fetch(a,r),c=i.localPath?{packageFs:new n.CwdFS(n.PortablePath.root),prefixPath:n.ppath.relative(n.PortablePath.root,i.localPath)}:i;i!==c&&i.releaseFs&&i.releaseFs();const l=c.packageFs,p=n.ppath.join(c.prefixPath,s);return await l.readFilePromise(p,"utf8")}const p={fetchers:[class{supports(e,t){return!!e.reference.startsWith("exec:")}getLocalPath(e,t){const{parentLocator:r,path:a}=o.structUtils.parseFileStyleRange(e.reference,{protocol:"exec:"});if(n.ppath.isAbsolute(a))return a;const s=t.fetcher.getLocalPath(r,t);return null===s?null:n.ppath.resolve(s,a)}async fetch(e,t){const r=t.checksums.get(e.locatorHash)||null,[a,n,s]=await t.cache.fetchPackageFromCache(e,r,{onHit:()=>t.report.reportCacheHit(e),onMiss:()=>t.report.reportCacheMiss(e),loader:()=>this.fetchFromDisk(e,t),skipIntegrityCheck:t.skipIntegrityCheck});return{packageFs:a,releaseFs:n,prefixPath:o.structUtils.getIdentVendorPath(e),localPath:this.getLocalPath(e,t),checksum:s}}async fetchFromDisk(e,t){const r=await l(e.reference,"exec:",t);return n.xfs.mktempPromise(async a=>{const s=n.ppath.join(a,"generator.js");return await n.xfs.writeFilePromise(s,r),n.xfs.mktempPromise(async r=>{if(await this.generatePackage(r,e,s,t),!n.xfs.existsSync(n.ppath.join(r,"build")))throw new Error("The script should have generated a build directory");return await o.tgzUtils.makeArchiveFromDirectory(n.ppath.join(r,"build"),{prefixPath:o.structUtils.getIdentVendorPath(e),compressionLevel:t.project.configuration.get("compressionLevel")})})})}async generatePackage(e,t,r,a){return await n.xfs.mktempPromise(async s=>{const i=await o.scriptUtils.makeScriptEnv({project:a.project,binFolder:s}),c=n.ppath.join(e,"runtime.js");return await n.xfs.mktempPromise(async a=>{const s=n.ppath.join(a,"buildfile.log"),l=n.xfs.createWriteStream(s),p=l,u=n.ppath.join(e,"generator"),h=n.ppath.join(e,"build");await n.xfs.mkdirPromise(u),await n.xfs.mkdirPromise(h);const f={tempDir:n.npath.fromPortablePath(u),buildDir:n.npath.fromPortablePath(h),locator:o.structUtils.stringifyLocator(t)};await n.xfs.writeFilePromise(c,`\n          // Expose 'Module' as a global variable\n          Object.defineProperty(global, 'Module', {\n            get: () => require('module'),\n            configurable: true,\n            enumerable: false,\n          });\n\n          // Expose non-hidden built-in modules as global variables\n          for (const name of Module.builtinModules.filter((name) => name !== 'module' && !name.startsWith('_'))) {\n            Object.defineProperty(global, name, {\n              get: () => require(name),\n              configurable: true,\n              enumerable: false,\n            });\n          }\n\n          // Expose the 'execEnv' global variable\n          Object.defineProperty(global, 'execEnv', {\n            value: {\n              ...${JSON.stringify(f)},\n            },\n            enumerable: true,\n          });\n        `);let d=i.NODE_OPTIONS||"";d=d.replace(/\s*--require\s+\S*\.pnp\.c?js\s*/g," ").trim(),i.NODE_OPTIONS=d,l.write(`# This file contains the result of Yarn generating a package (${o.structUtils.stringifyLocator(t)})\n`),l.write("\n");const{code:g}=await o.execUtils.pipevp(process.execPath,["--require",n.npath.fromPortablePath(c),n.npath.fromPortablePath(r),o.structUtils.stringifyIdent(t)],{cwd:e,env:i,stdin:null,stdout:l,stderr:p});if(0!==g)throw n.xfs.detachTemp(a),new Error(`Package generation failed (exit code ${g}, logs can be found here: ${s})`)})})}}],resolvers:[class{supportsDescriptor(e,t){return!!e.range.startsWith("exec:")}supportsLocator(e,t){return!!e.reference.startsWith("exec:")}shouldPersistResolution(e,t){return!1}bindDescriptor(e,t,r){return o.structUtils.bindDescriptor(e,{locator:o.structUtils.stringifyLocator(t)})}getResolutionDependencies(e,t){return[]}async getCandidates(e,t,r){if(!r.fetchOptions)throw new Error("Assertion failed: This resolver cannot be used unless a fetcher is configured");const{path:a,parentLocator:n}=s(e.range);if(null===n)throw new Error("Assertion failed: The descriptor should have been bound");const i=await l(o.structUtils.makeRange({protocol:"exec:",source:a,selector:a,params:{locator:o.structUtils.stringifyLocator(n)}}),"exec:",r.fetchOptions);return[c(e,{parentLocator:n,path:a,generatorHash:o.hashUtils.makeHash("1",i).slice(0,6),protocol:"exec:"})]}async getSatisfying(e,t,r){return null}async resolve(e,t){if(!t.fetchOptions)throw new Error("Assertion failed: This resolver cannot be used unless a fetcher is configured");const r=await t.fetchOptions.fetcher.fetch(e,t.fetchOptions),a=await o.miscUtils.releaseAfterUseAsync(async()=>await o.Manifest.find(r.prefixPath,{baseFs:r.packageFs}),r.releaseFs);return{...e,version:a.version||"0.0.0",languageName:t.project.configuration.get("defaultLanguageName"),linkType:o.LinkType.HARD,dependencies:a.dependencies,peerDependencies:a.peerDependencies,dependenciesMeta:a.dependenciesMeta,peerDependenciesMeta:a.peerDependenciesMeta,bin:a.bin}}}]}},966:e=>{e.exports=require("@yarnpkg/core")},688:e=>{e.exports=require("@yarnpkg/fslib")}},t={};function r(a){if(t[a])return t[a].exports;var o=t[a]={exports:{}};return e[a](o,o.exports,r),o.exports}return r.d=(e,t)=>{for(var a in t)r.o(t,a)&&!r.o(e,a)&&Object.defineProperty(e,a,{enumerable:!0,get:t[a]})},r.o=(e,t)=>Object.prototype.hasOwnProperty.call(e,t),r.r=e=>{"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},r(230)})();
return plugin;
}
};