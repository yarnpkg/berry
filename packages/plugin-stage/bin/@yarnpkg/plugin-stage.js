module.exports={name:"@yarnpkg/plugin-stage",factory:function(t){return function(t){var e={};function n(o){if(e[o])return e[o].exports;var a=e[o]={i:o,l:!1,exports:{}};return t[o].call(a.exports,a,a.exports,n),a.l=!0,a.exports}return n.m=t,n.c=e,n.d=function(t,e,o){n.o(t,e)||Object.defineProperty(t,e,{enumerable:!0,get:o})},n.r=function(t){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(t,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(t,"__esModule",{value:!0})},n.t=function(t,e){if(1&e&&(t=n(t)),8&e)return t;if(4&e&&"object"==typeof t&&t&&t.__esModule)return t;var o=Object.create(null);if(n.r(o),Object.defineProperty(o,"default",{enumerable:!0,value:t}),2&e&&"string"!=typeof t)for(var a in t)n.d(o,a,function(e){return t[e]}.bind(null,a));return o},n.n=function(t){var e=t&&t.__esModule?function(){return t.default}:function(){return t};return n.d(e,"a",e),e},n.o=function(t,e){return Object.prototype.hasOwnProperty.call(t,e)},n.p="",n(n.s=4)}([function(e,n){e.exports=t("@yarnpkg/fslib")},function(e,n){e.exports=t("@yarnpkg/core")},function(e,n){e.exports=t("clipanion")},function(e,n){e.exports=t("@yarnpkg/cli")},function(t,e,n){"use strict";n.r(e);function o(t,e,n,o){var a,i=arguments.length,r=i<3?e:null===o?o=Object.getOwnPropertyDescriptor(e,n):o;if("object"==typeof Reflect&&"function"==typeof Reflect.decorate)r=Reflect.decorate(t,e,n,o);else for(var s=t.length-1;s>=0;s--)(a=t[s])&&(r=(i<3?a(r):i>3?a(e,n,r):a(e,n))||r);return i>3&&r&&Object.defineProperty(e,n,r),r}var a,i=n(3),r=n(1),s=n(0),c=n(2);async function l(t,{marker:e}){do{if(s.xfs.existsSync(s.ppath.join(t,e)))return t;t=s.ppath.dirname(t)}while("/"!==t);return null}function u(t,e){let n=0,o=0;for(const a of t)"wip"!==a&&(e.test(a)?n+=1:o+=1);return n>=o}!function(t){t[t.CREATE=0]="CREATE",t[t.DELETE=1]="DELETE",t[t.ADD=2]="ADD",t[t.REMOVE=3]="REMOVE",t[t.MODIFY=4]="MODIFY"}(a||(a={}));const p=new Map([[a.CREATE,"create"],[a.DELETE,"delete"],[a.ADD,"add"],[a.REMOVE,"remove"],[a.MODIFY,"update"]]);const f=11;async function d(t){const{code:e,stdout:n}=await r.execUtils.execvp("git",["log","-1","--pretty=format:%H"],{cwd:t});return 0===e?n.trim():null}async function m(t,e){const n=[],o=e.filter(t=>"package.json"===s.ppath.basename(t.path));for(const{action:e,path:i}of o){const o=s.ppath.relative(t,i);if(e===a.MODIFY){const e=await d(t),{stdout:s}=await r.execUtils.execvp("git",["show",`${e}:${o}`],{cwd:t,strict:!0}),c=await r.Manifest.fromText(s),l=await r.Manifest.fromFile(i),u=new Map([...l.dependencies,...l.devDependencies]),p=new Map([...c.dependencies,...c.devDependencies]);for(const[t,e]of p){const o=r.structUtils.stringifyIdent(e),i=u.get(t);i?i.range!==e.range&&n.push([a.MODIFY,`${o} to ${i.range}`]):n.push([a.REMOVE,o])}for(const[t,e]of u)p.has(t)||n.push([a.ADD,r.structUtils.stringifyIdent(e)])}else if(e===a.CREATE){const t=await r.Manifest.fromFile(i);t.name?n.push([a.CREATE,r.structUtils.stringifyIdent(t.name)]):n.push([a.CREATE,"a package"])}else{if(e!==a.DELETE)throw new Error("Assertion failed: Unsupported action type");{const e=await d(t),{stdout:i}=await r.execUtils.execvp("git",["show",`${e}:${o}`],{cwd:t,strict:!0}),s=await r.Manifest.fromText(i);s.name?n.push([a.DELETE,r.structUtils.stringifyIdent(s.name)]):n.push([a.DELETE,"a package"])}}}const{code:i,stdout:c}=await r.execUtils.execvp("git",["log",`-${f}`,"--pretty=format:%s"],{cwd:t});return function(t,e){const n=function(t){return t.useComponent?"chore(yarn): ":""}(t),o=[],a=e.slice().sort((t,e)=>t[0]-e[0]);for(;a.length>0;){const[e,n]=a.shift();let i=p.get(e);t.useUpperCase&&0===o.length&&(i=`${i[0].toUpperCase()}${i.slice(1)}`),t.useThirdPerson&&(i+="s");const r=[n];for(;a.length>0&&a[0][0]===e;){const[,t]=a.shift();r.push(t)}r.sort();let s=r.shift();1===r.length?s+=" (and one other)":r.length>1&&(s+=` (and ${r.length} others)`),o.push(`${i} ${s}`)}return`${n}${o.join(", ")}`}(function(t){return{useThirdPerson:u(t,/^(\w\(\w+\):\s*)?\w+s/),useUpperCase:u(t,/^(\w\(\w+\):\s*)?[A-Z]/),useComponent:u(t,/^\w\(\w+\):/)}}(0===i?c.split(/\n/g).filter(t=>""!==t):[]),n)}const h=[{findRoot:async t=>await l(t,{marker:Object(s.toFilename)(".git")}),async filterChanges(t,e,n){const{stdout:o}=await r.execUtils.execvp("git",["status","-s"],{cwd:t,strict:!0}),i=o.toString().split(/\n/g);return[].concat(...i.map(e=>{if(""===e)return[];const n=e.slice(0,3),o=s.ppath.resolve(t,e.slice(3));return"?? "===n&&e.endsWith("/")?function(t){const e=[],n=[t];for(;n.length>0;){const t=n.pop(),o=s.xfs.readdirSync(t);for(const a of o){const o=s.ppath.resolve(t,a);s.xfs.lstatSync(o).isDirectory()?n.push(o):e.push(o)}}return e}(o).map(t=>({action:a.CREATE,path:t})):" A "===n||"?? "===n?[{action:a.CREATE,path:o}]:" M "===n?[{action:a.MODIFY,path:o}]:" D "===n?[{action:a.DELETE,path:o}]:[]})).filter(t=>(function(t,{roots:e,names:n}){if(n.has(s.ppath.basename(t)))return!0;do{if(e.has(t))return!0;t=s.ppath.dirname(t)}while("/"!==t);return!1})(t.path,{roots:e,names:n}))},genCommitMessage:async(t,e)=>await m(t,e),async makeCommit(t,e,n){const o=e.map(t=>s.npath.fromPortablePath(t.path));await r.execUtils.execvp("git",["add","-N","--",...o],{cwd:t,strict:!0}),await r.execUtils.execvp("git",["commit","-m",`${n}\n\nCommit generated via \`yarn stage\`\n`,"--",...o],{cwd:t,strict:!0})},async makeReset(t,e){const n=e.map(t=>s.npath.fromPortablePath(t.path));await r.execUtils.execvp("git",["reset","HEAD","--",...n],{cwd:t,strict:!0})}},{findRoot:async t=>await l(t,{marker:Object(s.toFilename)(".hg")}),filterChanges:async(t,e,n)=>[],genCommitMessage:async(t,e)=>"",async makeCommit(t,e,n){},async makeReset(t,e){},async makeUpdate(t,e){}}];class g extends i.BaseCommand{constructor(){super(...arguments),this.commit=!1,this.reset=!1,this.update=!1,this.dryRun=!1}async execute(){const t=await r.Configuration.find(this.context.cwd,this.context.plugins),{project:e}=await r.Project.find(t,this.context.cwd),{driver:n,root:o}=await async function(t){let e=null,n=null;for(const o of h)if(null!==(n=await o.findRoot(t))){e=o;break}if(null===e||null===n)throw new c.UsageError("No stage driver has been found for your current project");return{driver:e,root:n}}(e.cwd),a=[t.get("bstatePath"),t.get("cacheFolder"),t.get("globalFolder"),t.get("virtualFolder"),t.get("yarnPath")];await t.triggerHook(t=>t.populateYarnPaths,e,t=>{a.push(t)});const i=new Set;for(const t of a)for(const e of y(o,t))i.add(e);const l=new Set([t.get("rcFilename"),t.get("lockfileFilename"),"package.json"]),u=await n.filterChanges(o,i,l),p=await n.genCommitMessage(o,u);if(this.dryRun)if(this.commit)this.context.stdout.write(`${p}\n`);else for(const t of u)this.context.stdout.write(`${s.npath.fromPortablePath(t.path)}\n`);else 0===u.length?this.context.stdout.write("No changes found!"):this.commit?await n.makeCommit(o,u,p):this.reset&&await n.makeReset(o,u)}}function y(t,e){const n=[];if(null===e)return n;for(;;){let o;(e===t||e.startsWith(`${t}/`))&&n.push(e);try{o=s.xfs.statSync(e)}catch(t){break}if(!o.isSymbolicLink())break;e=s.ppath.resolve(s.ppath.dirname(e),s.xfs.readlinkSync(e))}return n}g.usage=c.Command.Usage({description:"add all yarn files to your vcs",details:"\n      This command will add to your staging area the files belonging to Yarn (typically any modified `package.json` and `.yarnrc.yml` files, but also linker-generated files, cache data, etc). It will take your ignore list into account, so the cache files won't be added if the cache is ignored in a `.gitignore` file (assuming you use Git).\n\n      Running `--reset` will instead remove them from the staging area (the changes will still be there, but won't be committed until you stage them back).\n\n      Since the staging area is a non-existent concept in Mercurial, Yarn will always create a new commit when running this command on Mercurial repositories. You can get this behavior when using Git by using the `--commit` flag which will directly create a commit.\n    ",examples:[["Adds all modified project files to the staging area","yarn stage"],["Creates a new commit containing all modified project files","yarn stage --commit"]]}),o([c.Command.Boolean("-c,--commit")],g.prototype,"commit",void 0),o([c.Command.Boolean("-r,--reset")],g.prototype,"reset",void 0),o([c.Command.Boolean("-u,--update")],g.prototype,"update",void 0),o([c.Command.Boolean("-n,--dry-run")],g.prototype,"dryRun",void 0),o([c.Command.Path("stage")],g.prototype,"execute",null);const w={commands:[g]};e.default=w}])}};