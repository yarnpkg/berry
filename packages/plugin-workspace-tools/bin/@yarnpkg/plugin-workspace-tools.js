module.exports={name:"@yarnpkg/plugin-workspace-tools",factory:function(e){return function(e){var t={};function o(n){if(t[n])return t[n].exports;var a=t[n]={i:n,l:!1,exports:{}};return e[n].call(a.exports,a,a.exports,o),a.l=!0,a.exports}return o.m=e,o.c=t,o.d=function(e,t,n){o.o(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:n})},o.r=function(e){"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},o.t=function(e,t){if(1&t&&(e=o(e)),8&t)return e;if(4&t&&"object"==typeof e&&e&&e.__esModule)return e;var n=Object.create(null);if(o.r(n),Object.defineProperty(n,"default",{enumerable:!0,value:e}),2&t&&"string"!=typeof e)for(var a in e)o.d(n,a,function(t){return e[t]}.bind(null,a));return n},o.n=function(e){var t=e&&e.__esModule?function(){return e.default}:function(){return e};return o.d(t,"a",t),t},o.o=function(e,t){return Object.prototype.hasOwnProperty.call(e,t)},o.p="",o(o.s=7)}([function(t,o){t.exports=e("@yarnpkg/core")},function(t,o){t.exports=e("clipanion")},function(t,o){t.exports=e("@yarnpkg/cli")},function(t,o){t.exports=e("yup")},function(t,o){t.exports=e("os")},function(e,t,o){"use strict";const n=o(6),a=e=>{if(e<1)throw new TypeError("Expected `concurrency` to be a number from 1 and up");const t=[];let o=0;const a=()=>{o--,t.length>0&&t.shift()()},r=(e,t,...r)=>{o++;const s=n(e,...r);t(s),s.then(a,a)},s=(n,...a)=>new Promise(s=>((n,a,...s)=>{o<e?r(n,a,...s):t.push(r.bind(null,n,a,...s))})(n,s,...a));return Object.defineProperties(s,{activeCount:{get:()=>o},pendingCount:{get:()=>t.length}}),s};e.exports=a,e.exports.default=a},function(e,t,o){"use strict";e.exports=(e,...t)=>new Promise(o=>{o(e(...t))})},function(e,t,o){"use strict";o.r(t);function n(e,t,o,n){var a,r=arguments.length,s=r<3?t:null===n?n=Object.getOwnPropertyDescriptor(t,o):n;if("object"==typeof Reflect&&"function"==typeof Reflect.decorate)s=Reflect.decorate(e,t,o,n);else for(var i=e.length-1;i>=0;i--)(a=e[i])&&(s=(r<3?a(s):r>3?a(t,o,s):a(t,o))||s);return r>3&&s&&Object.defineProperty(t,o,s),s}var a=o(2),r=o(0),s=o(1);class i extends a.BaseCommand{constructor(){super(...arguments),this.workspaces=[],this.json=!1,this.production=!1}async execute(){const e=await r.Configuration.find(this.context.cwd,this.context.plugins),{project:t,workspace:o}=await r.Project.find(e,this.context.cwd),n=await r.Cache.find(e);let s;if(0===this.workspaces.length){if(!o)throw new a.WorkspaceRequiredError(t.cwd,this.context.cwd);s=new Set([o])}else s=new Set(this.workspaces.map(e=>t.getWorkspaceByIdent(r.structUtils.parseIdent(e))));for(const e of s)for(const o of r.Manifest.hardDependencies)for(const n of e.manifest.getForScope(o).values()){const e=t.tryWorkspaceByDescriptor(n);null!==e&&s.add(e)}for(const e of t.workspaces)s.has(e)?this.production&&e.manifest.devDependencies.clear():(e.manifest.dependencies.clear(),e.manifest.devDependencies.clear(),e.manifest.peerDependencies.clear());return(await r.StreamReport.start({configuration:e,json:this.json,stdout:this.context.stdout,includeLogs:!0},async e=>{await t.install({cache:n,report:e,persistProject:!1}),await t.persistInstallStateFile()})).exitCode()}}i.usage=s.Command.Usage({category:"Workspace-related commands",description:"install a single workspace and its dependencies",details:"\n      This command will run an install as if the specified workspaces (and all other workspaces they depend on) were the only ones in the project. If no workspaces are explicitly listed, the active one will be assumed.\n\n      Note that this command is only very moderately useful when using zero-installs, since the cache will contain all the packages anyway - meaning that the only difference between a full install and a focused install would just be a few extra lines in the `.pnp.js` file, at the cost of introducing an extra complexity.\n\n      If the `--production` flag is set, only regular dependencies will be installed, and dev dependencies will be omitted.\n\n      If the `--json` flag is set the output will follow a JSON-stream output also known as NDJSON (https://github.com/ndjson/ndjson-spec).\n    "}),n([s.Command.Rest()],i.prototype,"workspaces",void 0),n([s.Command.Boolean("--json")],i.prototype,"json",void 0),n([s.Command.Boolean("--production")],i.prototype,"production",void 0),n([s.Command.Path("workspaces","focus")],i.prototype,"execute",null);var c=o(4),l=o(5),d=o.n(l),p=o(3);const u=(e,t)=>{const o=[];for(const n of e.workspacesCwds){const e=t.workspacesByCwd.get(n);e&&o.push(e,...u(e,t))}return o};class h extends a.BaseCommand{constructor(){super(...arguments),this.args=[],this.all=!1,this.verbose=!1,this.parallel=!1,this.interlaced=!1,this.topological=!1,this.topologicalDev=!1,this.include=[],this.exclude=[],this.private=!0}async execute(){const e=await r.Configuration.find(this.context.cwd,this.context.plugins),{project:t,workspace:o}=await r.Project.find(e,this.context.cwd);if(!this.all&&!o)throw new a.WorkspaceRequiredError(t.cwd,this.context.cwd);const n=this.cli.process([this.commandName,...this.args]),i=1===n.path.length&&"run"===n.path[0]&&void 0!==n.scriptName?n.scriptName:null;if(0===n.path.length)throw new s.UsageError("Invalid subcommand name for iteration - use the 'run' keyword if you wish to execute a script");const l=this.all?t.topLevelWorkspace:o,p=[l,...u(l,t)],h=[];for(const e of p)i&&!e.manifest.scripts.has(i)||i===process.env.npm_lifecycle_event&&e.cwd===o.cwd||this.include.length>0&&!this.include.includes(r.structUtils.stringifyIdent(e.locator))||this.exclude.length>0&&this.exclude.includes(r.structUtils.stringifyIdent(e.locator))||!1===this.private&&!0===e.manifest.private||h.push(e);let m=this.interlaced;this.parallel||(m=!0);const w=new Map,g=new Set,y=this.parallel?Math.max(1,Object(c.cpus)().length/2):1,b=d()(this.jobs||y);let v=0,x=null;const k=await r.StreamReport.start({configuration:e,stdout:this.context.stdout},async o=>{const n=async(t,{commandIndex:n})=>{!this.parallel&&this.verbose&&n>1&&o.reportSeparator();const a=function(e,{configuration:t,commandIndex:o,verbose:n}){if(!n)return null;const a=r.structUtils.convertToIdent(e.locator),s=`[${r.structUtils.stringifyIdent(a)}]:`,i=["#2E86AB","#A23B72","#F18F01","#C73E1D","#CCE2A3"],c=i[o%i.length];return t.format(s,c)}(t,{configuration:e,verbose:this.verbose,commandIndex:n}),[s,i]=f(o,{prefix:a,interlaced:m}),[c,l]=f(o,{prefix:a,interlaced:m});try{const e=await this.cli.run([this.commandName,...this.args],{cwd:t.cwd,stdout:s,stderr:c})||0;s.end(),c.end();const n=await i,r=await l;return this.verbose&&n&&r&&o.reportInfo(null,`${a} Process exited without output (exit code ${e})`),e}catch(e){throw s.end(),c.end(),await i,await l,e}};for(const e of h)w.set(e.anchoredLocator.locatorHash,e);for(;w.size>0&&!o.hasErrors();){const a=[];for(const[e,o]of w){if(g.has(o.anchoredDescriptor.descriptorHash))continue;let r=!0;if(this.topological||this.topologicalDev){const e=this.topologicalDev?new Map([...o.manifest.dependencies,...o.manifest.devDependencies]):o.manifest.dependencies;for(const o of e.values()){const e=t.tryWorkspaceByDescriptor(o);if(!(r=null===e||!w.has(e.anchoredLocator.locatorHash)))break}}if(r&&(g.add(o.anchoredDescriptor.descriptorHash),a.push(b(async()=>{const t=await n(o,{commandIndex:++v});return w.delete(e),g.delete(o.anchoredDescriptor.descriptorHash),t})),!this.parallel))break}if(0===a.length){const t=Array.from(w.values()).map(t=>r.structUtils.prettyLocator(e,t.anchoredLocator)).join(", ");return void o.reportError(r.MessageName.CYCLIC_DEPENDENCIES,`Dependency cycle detected (${t})`)}const s=(await Promise.all(a)).find(e=>0!==e);x=void 0!==s?1:x,(this.topological||this.topologicalDev)&&void 0!==s&&o.reportError(r.MessageName.UNNAMED,"The command failed for workspaces that are depended upon by other workspaces; can't satisfy the dependency graph")}});return null!==x?x:k.exitCode()}}function f(e,{prefix:t,interlaced:o}){const n=e.createStreamReporter(t),a=new r.miscUtils.DefaultStream;a.pipe(n,{end:!1}),a.on("finish",()=>{n.end()});const s=new Promise(e=>{n.on("finish",()=>{e(a.active)})});if(o)return[a,s];const i=new r.miscUtils.BufferStream;return i.pipe(a,{end:!1}),i.on("finish",()=>{a.end()}),[i,s]}h.schema=p.object().shape({jobs:p.number().min(2),parallel:p.boolean().when("jobs",{is:e=>e>1,then:p.boolean().oneOf([!0],"--parallel must be set when using --jobs"),otherwise:p.boolean()})}),h.usage=s.Command.Usage({category:"Workspace-related commands",description:"run a command on all workspaces",details:"\n      This command will run a given sub-command on current and all its descendant workspaces. Various flags can alter the exact behavior of the command:\n\n      - If `-p,--parallel` is set, the commands will be ran in parallel; they'll by default be limited to a number of parallel tasks roughly equal to half your core number, but that can be overridden via `-j,--jobs`.\n\n      - If `-p,--parallel` and `-i,--interlaced` are both set, Yarn will print the lines from the output as it receives them. If `-i,--interlaced` wasn't set, it would instead buffer the output from each process and print the resulting buffers only after their source processes have exited.\n\n      - If `-t,--topological` is set, Yarn will only run the command after all workspaces that depend on it through the `dependencies` field have successfully finished executing. If `--topological-dev` is set, both the `dependencies` and `devDependencies` fields will be considered when figuring out the wait points.\n\n      - If `--all` is set, Yarn will run the command on all the workspaces of a project. By default yarn runs the command only on current and all its descendant workspaces.\n\n      - The command may apply to only some workspaces through the use of `--include` which acts as a whitelist. The `--exclude` flag will do the opposite and will be a list of packages that mustn't execute the script.\n\n      Adding the `-v,--verbose` flag will cause Yarn to print more information; in particular the name of the workspace that generated the output will be printed at the front of each line.\n\n      If the command is `run` and the script being run does not exist the child workspace will be skipped without error.\n    ",examples:[["Publish current and all descendant packages","yarn workspaces foreach npm publish --tolerate-republish"],["Run build script on current and all descendant packages","yarn workspaces foreach run build"],["Run build script on current and all descendant packages in parallel, building dependent packages first","yarn workspaces foreach -pt run build"]]}),n([s.Command.String()],h.prototype,"commandName",void 0),n([s.Command.Proxy()],h.prototype,"args",void 0),n([s.Command.Boolean("-a,--all")],h.prototype,"all",void 0),n([s.Command.Boolean("-v,--verbose")],h.prototype,"verbose",void 0),n([s.Command.Boolean("-p,--parallel")],h.prototype,"parallel",void 0),n([s.Command.Boolean("-i,--interlaced")],h.prototype,"interlaced",void 0),n([s.Command.String("-j,--jobs")],h.prototype,"jobs",void 0),n([s.Command.Boolean("-t,--topological")],h.prototype,"topological",void 0),n([s.Command.Boolean("--topological-dev")],h.prototype,"topologicalDev",void 0),n([s.Command.Array("--include")],h.prototype,"include",void 0),n([s.Command.Array("--exclude")],h.prototype,"exclude",void 0),n([s.Command.Boolean("--private")],h.prototype,"private",void 0),n([s.Command.Path("workspaces","foreach")],h.prototype,"execute",null);const m={commands:[i,h]};t.default=m}])}};