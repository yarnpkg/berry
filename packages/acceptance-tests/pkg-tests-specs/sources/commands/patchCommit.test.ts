import {Filename, npath, ppath, xfs} from '@yarnpkg/fslib';

describe(`Commands`, () => {
  describe(`patch-commit`, () => {
    test(
      `it should generate a patch from a package folder`,
      makeTemporaryEnv({
        dependencies: {
          [`no-deps`]: `1.0.0`,
        },
      }, async ({path, run, source}) => {
        await run(`install`);

        const {stdout} = await run(`patch`, `no-deps`, `--json`);
        const {path: updateFolderN} = JSON.parse(stdout);

        const updateFolder = npath.toPortablePath(updateFolderN);
        const updateFile = ppath.join(updateFolder, `index.js` as Filename);

        const fileSource = await xfs.readFilePromise(updateFile, `utf8`);
        const fileUser = fileSource.replace(`require(dep)`, `42`);
        await xfs.writeFilePromise(updateFile, fileUser);

        await expect(run(`patch-commit`, updateFolder)).resolves.toMatchSnapshot();
      }),
    );

    test(
      `it should save patches into a local folder if requested`,
      makeTemporaryEnv({
        dependencies: {
          [`one-fixed-dep`]: `1.0.0`,
        },
      }, async ({path, run, source}) => {
        await run(`install`);

        const {stdout} = await run(`patch`, `one-fixed-dep`, `--json`);
        const {path: updateFolderN} = JSON.parse(stdout);

        const updateFolder = npath.toPortablePath(updateFolderN);
        const updateFile = ppath.join(updateFolder, `index.js` as Filename);

        const fileSource = await xfs.readFilePromise(updateFile, `utf8`);
        const fileUser = fileSource.replace(`require(dep)`, `42`);
        await xfs.writeFilePromise(updateFile, fileUser);

        await run(`patch-commit`, `-s`, npath.fromPortablePath(updateFolder));
        await run(`install`);

        await expect(source(`require('one-fixed-dep')`)).resolves.toMatchObject({
          dependencies: {
            [`no-deps`]: 42,
          },
        });
      }),
    );
  });
});
