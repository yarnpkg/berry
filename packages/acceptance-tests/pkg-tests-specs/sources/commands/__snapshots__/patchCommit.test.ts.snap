// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Commands patch-commit it should generate a patch from a package folder 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "diff --git a/index.js b/index.js
index a6bf8f586524db0daca72361135fe3fe0e64483b..aca59d18bee6854e40fd262a68733999e8fd44f5 100644
--- a/index.js
+++ b/index.js
@@ -5,6 +5,6 @@ module.exports = require(\`./package.json\`);
 for (const key of [\`dependencies\`, \`devDependencies\`, \`peerDependencies\`]) {
   for (const dep of Object.keys(module.exports[key] || {})) {
     // $FlowFixMe The whole point of this file is to be dynamic
-    module.exports[key][dep] = require(dep);
+    module.exports[key][dep] = 42;
   }
 }
",
}
`;
